src/sources/file.rs

```diff
@@ -600,28 +617,40 @@ fn wrap_with_line_agg(
     let logic = line_agg::Logic::new(config);
     Box::new(
         LineAgg::new(
-            rx.map(|line| (line.filename, line.text, (line.file_id, line.offset))),
+            rx.map(|line| {
+                (
+                    line.filename,
+                    line.text,
+                    (line.file_id, line.start_offset, line.end_offset),
+                )
+            }),
             logic,
         )
-        .map(|(filename, text, (file_id, offset))| Line {
-            text,
-            filename,
-            file_id,
-            offset,
-        }),
+        .map(
+            |(filename, text, (file_id, start_offset, end_offset))| Line {
+                text,
+                filename,
+                file_id,
+                start_offset,
+                end_offset,
+            },
+        ),
     )
 }
 
-fn create_event(
-    line: Bytes,
-    file: &str,
-    host_key: &str,
-    hostname: &Option<String>,
-    file_key: &Option<String>,
-) -> LogEvent {
+struct LogEventMetadata<'a> {
+    file: &'a str,
+    offset: u64,
+    host_key: &'a str,
+    hostname: &'a Option<String>,
+    file_key: &'a Option<String>,
+    offset_key: &'a Option<String>,
```

Comment:
I think these fields are going to be constant for every event, right? I think it would be best to separate them off so they can be set once and just reused.